Add images, fix typos (#486)
spatialos/UnrealGDK: SpatialGDK/Documentation/content/get-started/tutorial.md

### What will be covered?

In this tutorial you’ll implement cross server remote procedure calls (RPCs) in a simple third person shooter. The end result will be a multiplayer, cloud-hosted Unreal game running across multiple [server-workers]({{urlRoot}}/content/glossary#inspector) that players can seamlessly move between and shoot across. It will look something like this:

> ![]({{assetRoot}}assets/tutorial/cross-server.gif)

Reviewer: No more calling out the "like Unreal" claim?

Author: I removed this because I've called it out below. I can put it back in if you'd like though.

Reviewer: Yeah I think we can call it twice in one doc.

Author: okay, will reinstate :)

1. Navigate to `UnrealGDKThirdPersonShooter\Game` and create a `Plugins` directory.

> 2. In a terminal window,  change directory to the  `Plugins` directory and clone the [Unreal GDK](https://github.com/spatialos/UnrealGDK) repository using one of the following commands:

Reviewer 2: ```suggestion
1. In a terminal window, change directory to the `Plugins` directory and clone the [Unreal GDK](https://github.com/spatialos/UnrealGDK) repository using one of the following commands:
```

Author: outdated

1. In the same directory, open **ThirdPersonShooter.sln** with Visual Studio.
1. In the Solution Explorer window, right-click on **ThirdPersonShooter** and select **Build**.
1. Open **ThirdPersonShooter.uproject** in the Unreal Editor.

> 2. In the [GDK toolbar (SpatialOS documentation)]({{urlRoot}}/content/toolbars), click [`Schema` (SpatialOS documentation)]({{urlRoot}}/content/glossary#schema) to generate schema and then [`Snapshot` (SpatialOS documentation)]({{urlRoot}}/content/glossary#snapshot) to generate a snapshot.

Reviewer 2: ```suggestion
1. In the [GDK toolbar (SpatialOS documentation)]({{urlRoot}}/content/toolbars), click [`Schema` (SpatialOS documentation)]({{urlRoot}}/content/glossary#schema) to generate schema and then [`Snapshot` (SpatialOS documentation)]({{urlRoot}}/content/glossary#snapshot) to generate a snapshot.
```

Author: outdated

1. In the same directory, open **ThirdPersonShooter.sln** with Visual Studio.
1. In the Solution Explorer window, right-click on **ThirdPersonShooter** and select **Build**.
1. Open **ThirdPersonShooter.uproject** in the Unreal Editor.
1. In the [GDK toolbar (SpatialOS documentation)]({{urlRoot}}/content/toolbars), click [`Schema` (SpatialOS documentation)]({{urlRoot}}/content/glossary#schema) to generate schema and then [`Snapshot` (SpatialOS documentation)]({{urlRoot}}/content/glossary#snapshot) to generate a snapshot.

### Deploy the project locally

In this section you’ll run a [local deployment](https://docs.improbable.io/reference/latest/shared/glossary#local-deployment) of the project. As the name suggests, local deployments run on your development machine (you will [cloud deploy](https://docs.improbable.io/reference/latest/shared/glossary#cloud-deployment) later in this tutorial).

> 1. In a text editor, open `UnrealGDKThirdPersonShooter\spatial\default_launch.json` and set `rows` to `2`. This instructs SpatialOS that you will be launching two server-workers.

Reviewer 3: Provide the section here , maybe something like "under the load_balancing/layer_configurations/rectangle_grid." This is what the Json look like:
"load_balancing": {
    "layer_configurations": [
      {
          "layer": "UnrealWorker",
          "rectangle_grid": {

1. In the Unreal Editor, in the Unreal toolbar, open the **Play** drop-down menu.
1. Under **Multiplayer Options**, enter the number of players as **2**
1. Enter the number of servers as **2**
1. Check the box next to **Run Dedicated Server**

> 1. In Unreal Editor, in the SpatialOS GDK toolbar, select **Launch**. This is the green play icon, not to be confused with Unreal’s Launch button. This will open a terminal window and run the [`spatial local launch`](https://docs.improbable.io/reference/latest/shared/spatial-cli/spatial-local-launch#spatial-local-launch) command, which starts the [SpatialOS Runtime (SpatialOS documentation)](https://docs.improbable.io/reference/latest/shared/glossary#the-runtime). It's ready when you see `SpatialOS ready. Access the inspector at http://localhost:21000/inspector`.

Reviewer 3: **Note:** This is the green play iton, not...

    ```

    > **Note:** You only want to replicate the `CurrentHealth` variable to the client that owns this Actor, thus you specify the `COND_AutonomousOnly` flag.

    Finally, you need to implement the `OnRep_CurentHealth` function so that the player health UI gets updated when the `CurrentHealth` variable is replicated:

1. In your IDE, open `UnrealGDKThirdPersonShooter\Game\Source\ThirdPersonShooter\Characters\TPSCharacter.h`.
1. In the public scope of the class, insert the following snippet:
    
    ```	
    UFUNCTION()
    void OnRep_CurrentHealth();
    ```

1. In your IDE, open `UnrealGDKThirdPersonShooter\Game\Source\ThirdPersonShooter\Characters\TPSCharacter.cpp` and insert the following snippet:

    ```
    void ATPSCharacter::OnRep_CurrentHealth()
    {
    	if (GetNetMode() != NM_DedicatedServer)
    	{
		    ATPSPlayerController* PC = Cast<ATPSPlayerController>(GetController());
    		if (PC)
    		{
    			PC->UpdateHealthUI(CurrentHealth, MaxHealth);
    		}
    		else
    		{
    			UE_LOG(LogTPS, Warning, TEXT("Couldn't find a player controller for character: %s"), *this->GetName());
    		}
    	}
    }
    ```

> The workflow you just used mirrors that of native Unreal.

Reviewer: ```suggestion
Notice that the workflow you just used mirrors that of native Unreal.
```

> Because you’ve changed code in a function you now need build **ThirdPersonShooter.sln**, generate schema and a new snapshot. To do this:

Reviewer: ```suggestion
Because you’ve changed code in a function you now need build your project, generate schema and a new snapshot. To do this:
```

1. Open **ThirdPersonShooter.sln** with Visual Studio.
1. In the Solution Explorer window, right-click on **ThirdPersonShooter** and select **Build**.
1. Open **ThirdPersonShooter.uproject** in the Unreal Editor and click `Schema` and then `Snapshot`.

Now let’s test our health replication in another local deployment.

### Deploy the project locally (again)

1. In Unreal Editor, in the SpatialOS GDK toolbar, select **Launch**. It's ready when you see `SpatialOS ready. Access the inspector at [http://localhost:21000/inspector]()`.
1. From the Unreal Editor toolbar, click **Play** to run the game.

Notice that health now decrements when you are shot.

### View your SpatialOS world in the Inspector

![]({{assetRoot}}assets/tutorial/inspector.png)

> 

Reviewer 3: Let's learn the use the Inspector ->Let's learn to the use the Inspector

1. Under **Multiplayer Options**, enter the number of players as **2**
1. Enter the number of servers as **2**
1. Check the box next to **Run Dedicated Server**
1. In Unreal Editor, in the SpatialOS GDK toolbar, select **Launch**. This is the green play icon, not to be confused with Unreal’s Launch button. This will open a terminal window and run the [`spatial local launch`](https://docs.improbable.io/reference/latest/shared/spatial-cli/spatial-local-launch#spatial-local-launch) command, which starts the [SpatialOS Runtime (SpatialOS documentation)](https://docs.improbable.io/reference/latest/shared/glossary#the-runtime). It's ready when you see `SpatialOS ready. Access the inspector at http://localhost:21000/inspector`.
1. From the Unreal Editor toolbar, click **Play** to run the game. This starts two headless server-workers and two [client-workers](https://docs.improbable.io/reference/latest/shared/glossary#client-worker).

Notice that when players shoot each other, their health does not go down. It's not much fun with no skin in the game is it? Let’s fix the health system.

### Replicate health changes

> In this ThirdPersonShooter project each `TPSCharacter` contains a variable called `CurrentHealth`, which keeps track of that charachter's health. On your servers, `CurrentHealth` is reduced whenever a character is shot, but this reduction is not replicated on the clients connected to the game. This is because the `CurrentHealth` variable is not setup for replication.

Reviewer 2: ```suggestion
In this ThirdPersonShooter project each `TPSCharacter` contains a variable called `CurrentHealth`, which keeps track of that character's health. On your servers, `CurrentHealth` is reduced whenever a character is shot, but this reduction is not replicated on the clients connected to the game. This is because the `CurrentHealth` variable is not setup for replication.
```

1. Navigate to the `GetLifetimeReplicatedProps` function (which is implementd around line 182), and insert the following snippet:

    ```
    // Only replicate health to the owning client.
    DOREPLIFETIME_CONDITION(ATPSCharacter, CurrentHealth, COND_AutonomousOnly);
    ```

    > **Note:** You only want to replicate the `CurrentHealth` variable to the client that owns this Actor, thus you specify the `COND_AutonomousOnly` flag.

    Finally, you need to implement the `OnRep_CurentHealth` function so that the player health UI gets updated when the `CurrentHealth` variable is replicated:

1. In your IDE, open `UnrealGDKThirdPersonShooter\Game\Source\ThirdPersonShooter\Characters\TPSCharacter.h`.
1. In the public scope of the class, insert the following snippet:
    
    ```	
    UFUNCTION()
    void OnRep_CurrentHealth();
    ```

1. In your IDE, open `UnrealGDKThirdPersonShooter\Game\Source\ThirdPersonShooter\Characters\TPSCharacter.cpp` and insert the following snippet:

    ```
    void ATPSCharacter::OnRep_CurrentHealth()
    {
    	if (GetNetMode() != NM_DedicatedServer)
    	{

> 		    ATPSPlayerController* PC = Cast<ATPSPlayerController>(GetController());

Reviewer 2: ```suggestion
		ATPSPlayerController* PC = Cast<ATPSPlayerController>(GetController());
```

    		if (PC)
    		{
    			PC->UpdateHealthUI(CurrentHealth, MaxHealth);
    		}
    		else
    		{
    			UE_LOG(LogTPS, Warning, TEXT("Couldn't find a player controller for character: %s"), *this->GetName());
    		}
    	}
    }
    ```

Notice that the workflow you just used mirrors that of native Unreal.

> Because you’ve changed code in a function you now need build your project, generate schema and a new snapshot. To do this:

Reviewer 2: ```suggestion
Because you’ve changed code in a function you now need to build your project, generate schema and a new snapshot. To do this:
```

1. From the Unreal Editor toolbar, click **Play** to run the game.

Notice that health now decrements when you are shot.

### View your SpatialOS world in the Inspector

![]({{assetRoot}}assets/tutorial/inspector.png)

[The Inspector]({{urlRoot}}/content/glossary#inspector) provides a real-time view of what is happening in your [SpatialOS world]({{urlRoot}}/content/glossary#game-world). It’s a powerful tool for monitoring and debugging both during development and when your game is live in production. Let’s use the Inspector to visualise the areas that each of our server-workers have [authority]({{urlRoot}}/content/glossary#authority) (that is, read and write access) over.

1. Access the inspector at [http://localhost:21000/inspector](http://localhost:21000/inspector).
1. In the **View** tab, click the checkboxes next to both of the **UnrealWorkers**. This will cause the Inspector to display the areas that the server-workers have authority over as two coloured zones.
1. Back in your two Unreal game clients, run around and shoot.

> 1. Using the Inspector to track the location of your two players, notice that if you position them in the area of authority then their shots damage each other, but if they are on different servers, they can’t damage each other. Let’s fix that.

Reviewer 2: ```suggestion
1. Using the Inspector to track the location of your two players, notice that if you position them in the same area of authority then their shots damage each other, but if they are on different servers, they can’t damage each other. Let’s fix that.
```

1. In your IDE, open `UnrealGDKThirdPersonShooter\Game\Source\ThirdPersonShooter\Characters\TPSCharacter.h`.
1. On line 74, add this snippet:

    ```
    UFUNCTION(CrossServer, Reliable)
    void TakeGunDamageCrossServer(float Damage, const struct FDamageEvent& DamageEvent, AController* EventInstigator, AActor* DamageCauser);
    ```

>     This snippet creates a new `UFUNCTION` marked with the function tags [CrossServer]({{urlRoot}}/content/cross-server-rpcs) and [Reliable (Unreal documentation)](https://wiki.unrealengine.com/Replication#Reliable_vs_Unreliable_Function_Call_Replication). The CrossServer tag forces this function to be executed as as cross-server RPC.

Reviewer 2: ```suggestion
    This snippet creates a new `UFUNCTION` marked with the function tags [CrossServer]({{urlRoot}}/content/cross-server-rpcs) and [Reliable (Unreal documentation)](https://wiki.unrealengine.com/Replication#Reliable_vs_Unreliable_Function_Call_Replication). The CrossServer tag forces this function to be executed as a cross-server RPC.
```

Author: outdated

> 3. In your IDE, open `UnrealGDKThirdPersonShooter\Game\Source\ThirdPersonShooter\Characters\TPSCharacter.cpp`.

Reviewer 2: ```suggestion
1. In your IDE, open `UnrealGDKThirdPersonShooter\Game\Source\ThirdPersonShooter\Characters\TPSCharacter.cpp`.
```

Author: outdated

> 4. Replace the TakeDamage function (lines 514-548) with this snippet:

Reviewer 2: ```suggestion
1. Replace the TakeDamage function (lines 514-548) with this snippet:
```

Author: outdated, the slightly odd formatting is necessitated by improbadoc

1. In your IDE, open `UnrealGDKThirdPersonShooter\Game\Source\ThirdPersonShooter\Characters\TPSCharacter.h`.
1. On line 74, add this snippet:

    ```
    UFUNCTION(CrossServer, Reliable)
    void TakeGunDamageCrossServer(float Damage, const struct FDamageEvent& DamageEvent, AController* EventInstigator, AActor* DamageCauser);
    ```

>     This snippet creates a new `UFUNCTION` marked with the function tags [CrossServer]({{urlRoot}}/content/cross-server-rpcs) and [Reliable (Unreal documentation)](https://wiki.unrealengine.com/Replication#Reliable_vs_Unreliable_Function_Call_Replication). The CrossServer tag forces this function to be executed as as cross-server RPC.

Reviewer 2: ```suggestion
    This snippet creates a new `UFUNCTION` marked with the function tags [CrossServer]({{urlRoot}}/content/cross-server-rpcs) and [Reliable (Unreal documentation)](https://wiki.unrealengine.com/Replication#Reliable_vs_Unreliable_Function_Call_Replication). The CrossServer tag forces this function to be executed as a cross-server RPC.
```

1. In your IDE, open `UnrealGDKThirdPersonShooter\Game\Source\ThirdPersonShooter\Characters\TPSCharacter.cpp`.

> 1. Replace the TakeDamage function (lines 514-548) with this snippet:

Reviewer 2: ```suggestion
1. Replace the `TakeDamage` function (lines 514-548) with this snippet:
```

### Play your game

![]({{assetRoot}}assets/tutorial/console.png)

When your deployment has launched, SpatialOS automatically opens the [Console](https://console.improbable.io) in your browser.

1. In the Console, Select the **Launch** button on the left of the page, and then click the **Launch** button that appears in the centre of the page. The SpatialOS Launcher, which was installed along with SpatialOS, downloads the game client for this deployment and runs it on your local machine.
![]({{assetRoot}}assets/tutorial/launch.png)

> 2. Once the client has launched, enter the game and fire a few celebratory shots - you are now playing in your first SpatialOS cloud deployment!

Reviewer 2: ```suggestion
1. Once the client has launched, enter the game and fire a few celebratory shots - you are now playing in your first SpatialOS cloud deployment!
```

Author: outdated

### Invite your friends

1. To invite other players to this game, head back to the Deployment Overview page in your [Console](https://console.improbable.io), and select the **Share button**.

> 2. Share the generated link with your friends.

Reviewer 2: ```suggestion
1. Share the generated link with your friends!
```

Author: outdated
